lib/grammar.js

```diff
@@ -367,118 +394,143 @@ var grammar = module.exports = {
+      format: 'max-message-size:%s',
     },
     {
       // RFC7273
       // a=ts-refclk:ptp=IEEE1588-2008:39-A7-94-FF-FE-07-CB-D0:37
-      push:'tsRefClocks',
+      push: 'tsRefClocks',
       reg: /^ts-refclk:([^\s=]*)(?:=(\S*))?/,
       names: ['clksrc', 'clksrcExt'],
       format: function (o) {
         return 'ts-refclk:%s' + (o.clksrcExt != null ? '=%s' : '');
-      }
+      },
     },
     {
       // RFC7273
       // a=mediaclk:direct=963214424
-      name:'mediaClk',
+      name: 'mediaClk',
       reg: /^mediaclk:(?:id=(\S*))? *([^\s=]*)(?:=(\S*))?(?: *rate=(\d+)\/(\d+))?/,
-      names: ['id', 'mediaClockName', 'mediaClockValue', 'rateNumerator', 'rateDenominator'],
+      names: [
+        'id',
+        'mediaClockName',
+        'mediaClockValue',
+        'rateNumerator',
+        'rateDenominator',
+      ],
       format: function (o) {
         var str = 'mediaclk:';
-        str += (o.id != null ? 'id=%s %s' : '%v%s');
-        str += (o.mediaClockValue != null ? '=%s' : '');
-        str += (o.rateNumerator != null ? ' rate=%s' : '');
-        str += (o.rateDenominator != null ? '/%s' : '');
+        str += o.id != null ? 'id=%s %s' : '%v%s';
+        str += o.mediaClockValue != null ? '=%s' : '';
+        str += o.rateNumerator != null ? ' rate=%s' : '';
+        str += o.rateDenominator != null ? '/%s' : '';
         return str;
-      }
+      },
     },
     {
       // a=keywds:keywords
       name: 'keywords',
       reg: /^keywds:(.+)$/,
-      format: 'keywds:%s'
+      format: 'keywds:%s',
     },
     {
       // a=content:main
       name: 'content',
       reg: /^content:(.+)/,
-      format: 'content:%s'
+      format: 'content:%s',
     },
     // BFCP https://tools.ietf.org/html/rfc4583
     {
       // a=floorctrl:c-s
       name: 'bfcpFloorCtrl',
       reg: /^floorctrl:(c-only|s-only|c-s)/,
-      format: 'floorctrl:%s'
+      format: 'floorctrl:%s',
     },
     {
       // a=confid:1
       name: 'bfcpConfId',
       reg: /^confid:(\d+)/,
-      format: 'confid:%s'
+      format: 'confid:%s',
     },
     {
       // a=userid:1
       name: 'bfcpUserId',
       reg: /^userid:(\d+)/,
-      format: 'userid:%s'
+      format: 'userid:%s',
     },
     {
       // a=floorid:1
       name: 'bfcpFloorId',
       reg: /^floorid:(.+) (?:m-stream|mstrm):(.+)/,
       names: ['id', 'mStream'],
-      format: 'floorid:%s mstrm:%s'
+      format: 'floorid:%s mstrm:%s',
+    },
+    {
+      // RFC 3312 precondition
+      push: 'precondition',
+      reg: /^(curr|des|conf):(qos|\S*)(?: (mandatory|optional|none|failure|unknown))? (e2e|local|remote) (none|sendrecv|send|recv)/,
+      names: ['state', 'type', 'strength', 'status', 'direction'],
+      format: function (o) {
+        var str = '%s:%s ';
+        str += o.strength != null ? '%s ' : '%v';
+        str += o.status != null ? '%s ' : '';
+        str += o.direction != null ? '%s ' : '';
+        return str;
+      },
```

Comment:
This seems to be the part of the thing that you are actually adding. From a quick scan, looks sensible.
Are you able to add some tests for it?

Reply:
Hi, I tried but not sure why tests are not runnable. I'll spend more time.